File: airbyte-notification/src/main/java/io/airbyte/notification/messages/SyncSummary.java
/*
 * Copyright (c) 2020-2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.notification.messages;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@Builder
public class SyncSummary {

  private WorkspaceInfo workspace;

  private ConnectionInfo connection;

  private SourceInfo source;

  private DestinationInfo destination;

  private Long jobId;

  private boolean isSuccess;

  private Instant startedAt;

  private Instant finishedAt;

  private long bytesEmitted;

  private long bytesCommitted;

  private long recordsEmitted;

  private long recordsCommitted;

  private long recordsFilteredOut;

  private long bytesFilteredOut;

  private String errorMessage;

  private static String formatVolume(final long bytes) {
    long currentValue = bytes;
    for (String unit : List.of("B", "kB", "MB", "GB")) {
      var byteLimit = 1024;
      if (currentValue < byteLimit) {
        return String.format("%d %s", currentValue, unit);
      }
      currentValue = currentValue / byteLimit;
    }
    return String.format("%d TB", currentValue);
  }

  private static String formatDuration(final Instant start, final Instant end) {
    Duration duration = Duration.between(start, end);
    if (duration.toMinutes() == 0) {
      return String.format("%d sec", duration.toSecondsPart());
    } else if (duration.toHours() == 0) {
      return String.format("%d min %d sec", duration.toMinutesPart(), duration.toSecondsPart());
    } else if (duration.toDays() == 0) {
      return String.format("%d hours %d min", duration.toHoursPart(), duration.toMinutesPart());
    }
    return String.format("%d days %d hours", duration.toDays(), duration.toHoursPart());
  }

  public Long getDurationInSeconds() {
    if (startedAt != null && finishedAt != null) {
      return Duration.between(startedAt, finishedAt).getSeconds();
    }
    return null;
  }

  public String getDurationFormatted() {
    if (startedAt != null && finishedAt != null) {
      return formatDuration(startedAt, finishedAt);
    }
    return null;
  }

  public String getBytesEmittedFormatted() {
    return formatVolume(bytesEmitted);
  }

  public String getBytesCommittedFormatted() {
    return formatVolume(bytesCommitted);
  }

}

File: airbyte-notification/src/main/kotlin/io/airbyte/notification/messages/SyncSummary.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.notification.messages

import io.airbyte.config.FailureReason
import java.time.Duration
import java.time.Instant

data class SyncSummary(
  val workspace: WorkspaceInfo,
  val connection: ConnectionInfo,
  val source: SourceInfo,
  val destination: DestinationInfo,
  val jobId: Long,
  val isSuccess: Boolean,
  val startedAt: Instant?,
  val finishedAt: Instant?,
  val bytesEmitted: Long,
  val bytesCommitted: Long,
  val recordsEmitted: Long,
  val recordsCommitted: Long,
  val recordsFilteredOut: Long = 0,
  val bytesFilteredOut: Long = 0,
  // Note: It's a little odd to rename what we call "failure" internally to "error" externally, but we already locked in this terminology and it doesn't feel like it warrants a breaking change.
  val errorMessage: String? = null,
  val errorType: FailureReason.FailureType? = null,
  val errorOrigin: FailureReason.FailureOrigin? = null,
) {
  fun getDurationInSeconds(): Long? =
    when {
      startedAt != null && finishedAt != null -> Duration.between(startedAt, finishedAt).seconds
      else -> null
    }

  fun getDurationFormatted(): String? =
    when {
      startedAt != null && finishedAt != null -> formatDuration(startedAt, finishedAt)
      else -> null
    }

  fun getBytesEmittedFormatted() = formatVolume(bytesEmitted)

  fun getBytesCommittedFormatted() = formatVolume(bytesCommitted)
}

private fun formatDuration(
  start: Instant,
  end: Instant,
): String {
  val duration = Duration.between(start, end)
  return when {
    duration.toMinutes() == 0L -> "${duration.toSecondsPart()} sec"
    duration.toHours() == 0L -> "${duration.toMinutesPart()} min ${duration.toSecondsPart()} sec"
    duration.toDays() == 0L -> "${duration.toHoursPart()} hours ${duration.toMinutesPart()} min"
    else -> "${duration.toDays()} days ${duration.toHours()} hours"
  }
}

private fun formatVolume(bytes: Long): String {
  var currentValue = bytes
  val byteLimit = 1024
  listOf("B", "kB", "MB", "GB").forEach { unit ->
    if (currentValue < byteLimit) {
      return "$currentValue $unit"
    }
    currentValue /= byteLimit
  }

  return "$currentValue TB"
}
